packages/melody-hooks/src/hooks/useReducer.js

```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright 2018 trivago N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState } from './useState';
+import { useCallback } from './useCallback';
+
+export const useReducer = (reducer, initialState) => {
+    if (typeof reducer !== 'function') {
+        throw new Error(
+            '`useReducer` expects a reducer function as first argument'
+        );
+    }
+    const [state, setState] = useState(initialState);
+    const dispatch = useCallback(action => {
+        const nextState = reducer(state, action);
```

Comment:
I think we could use our new "functional" `setState`. In this way the reducer always gets the most up-to-date state.
```js
const dispatch = useCallback(action => {
    setState(prevState => reducer(prevState, action));
});
```

Reply:
true, fixed :)